pkg/sql/copy_file_upload.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/blobs"
+	"github.com/cockroachdb/cockroach/pkg/internal/client"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/lib/pq"
+)
+
+const (
+	fileUploadTable = "crdb_internal.file_upload"
+	copyOptionDest  = "destination"
+)
+
+var copyFileOptionExpectValues = map[string]KVStringOptValidate{
+	copyOptionDest: KVStringOptRequireValue,
+}
+
+type copyError struct {
+	err error
+}
+
+var _ copyMachineInterface = &fileUploadMachine{}
+
+type fileUploadMachine struct {
+	c           *copyMachine
+	writeToFile *io.PipeWriter
+	writeError  chan copyError
+}
+
+func newFileUploadMachine(
+	conn pgwirebase.Conn,
+	n *tree.CopyFrom,
+	execCfg *ExecutorConfig,
+	resetPlanner func(p *planner, txn *client.Txn, txnTS time.Time, stmtTS time.Time),
+) (f *fileUploadMachine, retErr error) {
+	c := &copyMachine{
+		conn:    conn,
+		table:   &n.Table,
+		columns: n.Columns,
+		// The planner will be prepared before use.
+		p:            planner{execCfg: execCfg},
+		resetPlanner: resetPlanner,
+	}
+	f = &fileUploadMachine{c: c}
+
+	optsFn, err := f.c.p.TypeAsStringOpts(n.Options, copyFileOptionExpectValues)
+	if err != nil {
+		return nil, err
+	}
+	opts, err := optsFn()
+	if err != nil {
+		return nil, err
+	}
+
+	pr, pw := io.Pipe()
+	localStorage, err := blobs.NewLocalStorage(c.p.execCfg.Settings.ExternalIODir)
+	if err != nil {
+		return nil, err
+	}
+	// Check that the file does not already exist
+	_, err = localStorage.Stat(opts[copyOptionDest])
+	if err == nil {
+		return nil, fmt.Errorf("destination file already exists for %s", opts[copyOptionDest])
+	}
+	f.writeError = make(chan copyError)
```

Comment:
This chan needs to be size 1 otherwise it could leak goroutines. The happy case below is that localStorage.WriteFile completes successfully and attempts to send it over the chan. Once something receives over the chan, it closes the channel and exits the goroutine. Yay.

In fileUploadMachine.run, a chan receive is only done at the very end of the function. But there are two other returns. If either of those is taken, then this goroutine will block forever waiting to send on the chan, but the only possible receiver already exited.

If you instead make this chan buffered with length 1, then this goroutine is guaranteed to always be able to send on the chan, which will allow for both the go routine and the chan to be cleaned up in either success or error conditions.

Reply:
Thank you for clearing this up for me! I'm new to channels and found it a little hard to understand and research. This mentioned some keywords that helped a lot, along with the behaviors in different cases. Making those changes!